Test Failure: com.ibm.ws.ejbcontainer.session.passivation.tests.StatefulTimeoutTest.testXMLOnly_EE8_FEATURES

testXMLOnly_EE8_FEATURES:junit.framework.AssertionFailedError: 2020-11-30-19:48:51:577 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testXMLOnly on servlet com.ibm.ws.ejbcontainer.session.passivation.statefulTimeout.web.StatefulTimeoutServlet
java.lang.AssertionError: Timeout was 15000 ms, but bean timed out after sleeping 10000 ms.
    at com.ibm.ws.ejbcontainer.session.passivation.statefulTimeout.web.StatefulTimeoutServlet.testXMLOnly(StatefulTimeoutServlet.java:208)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at componenttest.app.FATServlet.doGet(FATServlet.java:71)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1257)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:745)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:442)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1226)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1010)
    at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
    at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:83)
    at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:936)
    at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1141)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:422)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:381)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:565)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:499)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:359)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:326)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
    at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:239)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
". expected:<0> but was:<1>
    at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:537)
    at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:495)
    at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:472)
    at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:444)
    at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:414)
    at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:83)
    at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:40)
    at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:197)
    at componenttest.rules.repeater.RepeatTests$CompositeRepeatTestActionStatement.evaluate(RepeatTests.java:115)
    at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:318)
    at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:171)

Further analysis is available at https://libh-proxy1.fyre.ibm.com/cognitive/defectAnalysis.html?failureGroupId=failure_group-e7e85d5c-06d7-455f-acc9-34f139b446ed

--- a/dev/com.ibm.ws.ejbcontainer.session.passivation_fat/test-applications/StatefulTimeoutWeb.war/src/com/ibm/ws/ejbcontainer/session/passivation/statefulTimeout/web/StatefulTimeoutServlet.java
+++ b/dev/com.ibm.ws.ejbcontainer.session.passivation_fat/test-applications/StatefulTimeoutWeb.war/src/com/ibm/ws/ejbcontainer/session/passivation/statefulTimeout/web/StatefulTimeoutServlet.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2009, 2018 IBM Corporation and others.
+ * Copyright (c) 2009, 2020 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -144,18 +144,30 @@ public class StatefulTimeoutServlet extends FATServlet {
         if (!recLogInit)
             initRecoveryLog();
 
-        TimeoutInExtFileOnlyBean bean = (TimeoutInExtFileOnlyBean) new InitialContext().lookup(JNDI_EXT_FILE_ONLY);
-        long invokeTime = bean.getInvocationTime();
-        FATHelper.sleep(BUFFER);
-        try {
-            invokeTime = bean.getInvocationTime();
-        } catch (NoSuchEJBException ex) {
-            fail("Bean timed out after " + BUFFER + " ms, when the ibm-ejb-jar-ext.xml file specified a 15 second timeout.");
+        int retry = 0;
+        long secondInvokeTime = 0;
+        TimeoutInExtFileOnlyBean bean = null;
+        while (bean == null) {
+            bean = (TimeoutInExtFileOnlyBean) new InitialContext().lookup(JNDI_EXT_FILE_ONLY);
+            long firstInvokeTime = bean.getInvocationTime();
+            FATHelper.sleep(BUFFER);
+            try {
+                secondInvokeTime = bean.getInvocationTime();
+            } catch (NoSuchEJBException ex) {
+                long elapsedTime = System.currentTimeMillis() - firstInvokeTime;
+                if ((retry < 3) && (elapsedTime > USER_TIMEOUT)) {
+                    svLogger.info("Retry - Slow system; elapsed time to call method > timeout : " + elapsedTime + " > " + USER_TIMEOUT);
+                    retry++;
+                    bean = null;
+                } else {
+                    fail("Bean timed out after " + BUFFER + " ms, when the ibm-ejb-jar-ext.xml file specified a 15 second timeout.");
+                }
+            }
         }
 
         FATHelper.sleep(USER_TIMEOUT + BUFFER);
         try {
-            long elapsedTime = bean.getInvocationTime() - invokeTime;
+            long elapsedTime = bean.getInvocationTime() - secondInvokeTime;
             fail("Timeout should have occurred after 15 seconds (specified in ibm-ejb-jar-ext.xml file) but did not timeout after waiting " + elapsedTime + " ms.");
         } catch (NoSuchEJBException ex) {
             svLogger.info("Caught expected NoSuchEJBException - PASS");
@@ -172,18 +184,30 @@ public class StatefulTimeoutServlet extends FATServlet {
         if (!recLogInit)
             initRecoveryLog();
 
-        StatefulTimeoutAnnotationOnlyBean bean = (StatefulTimeoutAnnotationOnlyBean) new InitialContext().lookup(JNDI_ANN_ONLY);
-        long invokeTime = bean.getInvocationTime();
-        FATHelper.sleep(BUFFER);
-        try {
-            invokeTime = bean.getInvocationTime();
-        } catch (NoSuchEJBException ex) {
-            fail("Timeout was " + USER_TIMEOUT + " ms, but bean timed out after sleeping " + BUFFER + " ms.");
+        int retry = 0;
+        long secondInvokeTime = 0;
+        StatefulTimeoutAnnotationOnlyBean bean = null;
+        while (bean == null) {
+            bean = (StatefulTimeoutAnnotationOnlyBean) new InitialContext().lookup(JNDI_ANN_ONLY);
+            long firstInvokeTime = bean.getInvocationTime();
+            FATHelper.sleep(BUFFER);
+            try {
+                secondInvokeTime = bean.getInvocationTime();
+            } catch (NoSuchEJBException ex) {
+                long elapsedTime = System.currentTimeMillis() - firstInvokeTime;
+                if ((retry < 3) && (elapsedTime > USER_TIMEOUT)) {
+                    svLogger.info("Retry - Slow system; elapsed time to call method > timeout : " + elapsedTime + " > " + USER_TIMEOUT);
+                    retry++;
+                    bean = null;
+                } else {
+                    fail("Timeout was " + USER_TIMEOUT + " ms, but bean timed out after sleeping " + BUFFER + " ms.");
+                }
+            }
         }
 
         FATHelper.sleep(USER_TIMEOUT + BUFFER);
         try {
-            long elapsedTime = bean.getInvocationTime() - invokeTime;
+            long elapsedTime = bean.getInvocationTime() - secondInvokeTime;
             fail("Timeout was " + USER_TIMEOUT + " ms, but bean did not time out after sleeping " + elapsedTime + " ms.");
         } catch (NoSuchEJBException ex) {
             svLogger.info("Caught expected NoSuchEJBException - PASS");
@@ -199,18 +223,30 @@ public class StatefulTimeoutServlet extends FATServlet {
         if (!recLogInit)
             initRecoveryLog();
 
-        StatefulTimeoutXMLOnlyBean bean = (StatefulTimeoutXMLOnlyBean) new InitialContext().lookup(JNDI_XML_ONLY);
-        long invokeTime = bean.getInvocationTime();
-        FATHelper.sleep(BUFFER);
-        try {
-            invokeTime = bean.getInvocationTime();
-        } catch (NoSuchEJBException ex) {
-            fail("Timeout was " + USER_TIMEOUT + " ms, but bean timed out after sleeping " + BUFFER + " ms.");
+        int retry = 0;
+        long secondInvokeTime = 0;
+        StatefulTimeoutXMLOnlyBean bean = null;
+        while (bean == null) {
+            bean = (StatefulTimeoutXMLOnlyBean) new InitialContext().lookup(JNDI_XML_ONLY);
+            long firstInvokeTime = bean.getInvocationTime();
+            FATHelper.sleep(BUFFER);
+            try {
+                secondInvokeTime = bean.getInvocationTime();
+            } catch (NoSuchEJBException ex) {
+                long elapsedTime = System.currentTimeMillis() - firstInvokeTime;
+                if ((retry < 3) && (elapsedTime > USER_TIMEOUT)) {
+                    svLogger.info("Retry - Slow system; elapsed time to call method > timeout : " + elapsedTime + " > " + USER_TIMEOUT);
+                    retry++;
+                    bean = null;
+                } else {
+                    fail("Timeout was " + USER_TIMEOUT + " ms, but bean timed out after sleeping " + BUFFER + " ms.");
+                }
+            }
         }
 
         FATHelper.sleep(USER_TIMEOUT + BUFFER);
         try {
-            long elapsedTime = bean.getInvocationTime() - invokeTime;
+            long elapsedTime = bean.getInvocationTime() - secondInvokeTime;
             fail("Timeout was " + USER_TIMEOUT + " ms, but bean did not time out after sleeping " + elapsedTime + " ms.");
         } catch (NoSuchEJBException ex) {
             svLogger.info("Caught expected NoSuchEJBException - PASS");
@@ -484,20 +520,32 @@ public class StatefulTimeoutServlet extends FATServlet {
         if (!recLogInit)
             initRecoveryLog();
 
-        StatefulTimeoutAnnotationOverrideExtFileBean bean = (StatefulTimeoutAnnotationOverrideExtFileBean) new InitialContext().lookup(JNDI_ANN_OVERRIDE_EXT_FILE);
-        // timeout specified in ibm-ejb-jar-ext.xml is 60 seconds
-        // timeout specified in annotation is 15 seconds
-        long invokeTime = bean.getInvocationTime();
-        FATHelper.sleep(BUFFER);
-        try {
-            invokeTime = bean.getInvocationTime();
-        } catch (NoSuchEJBException ex) {
-            fail("Expected timeout to be 15 seconds, but bean timed out after " + BUFFER + " ms.");
+        int retry = 0;
+        long secondInvokeTime = 0;
+        StatefulTimeoutAnnotationOverrideExtFileBean bean = null;
+        while (bean == null) {
+            bean = (StatefulTimeoutAnnotationOverrideExtFileBean) new InitialContext().lookup(JNDI_ANN_OVERRIDE_EXT_FILE);
+            // timeout specified in ibm-ejb-jar-ext.xml is 60 seconds
+            // timeout specified in annotation is 15 seconds
+            long firstInvokeTime = bean.getInvocationTime();
+            FATHelper.sleep(BUFFER);
+            try {
+                secondInvokeTime = bean.getInvocationTime();
+            } catch (NoSuchEJBException ex) {
+                long elapsedTime = System.currentTimeMillis() - firstInvokeTime;
+                if ((retry < 3) && (elapsedTime > USER_TIMEOUT)) {
+                    svLogger.info("Retry - Slow system; elapsed time to call method > timeout : " + elapsedTime + " > " + USER_TIMEOUT);
+                    retry++;
+                    bean = null;
+                } else {
+                    fail("Expected timeout to be 15 seconds, but bean timed out after " + BUFFER + " ms.");
+                }
+            }
         }
 
         FATHelper.sleep(USER_TIMEOUT + BUFFER);
         try {
-            long elapsedTime = bean.getInvocationTime() - invokeTime;
+            long elapsedTime = bean.getInvocationTime() - secondInvokeTime;
             fail("Timeout was " + USER_TIMEOUT + " ms, but bean did not time out after sleeping " + elapsedTime + " ms.");
         } catch (NoSuchEJBException ex) {
             svLogger.info("Caught expected NoSuchEJBException - PASS");
@@ -515,20 +563,32 @@ public class StatefulTimeoutServlet extends FATServlet {
         if (!recLogInit)
             initRecoveryLog();
 
-        StatefulTimeoutXMLOverrideExtFileBean bean = (StatefulTimeoutXMLOverrideExtFileBean) new InitialContext().lookup(JNDI_XML_OVERRIDE_EXT_FILE);
-        // timeout specified in ibm-ejb-jar-ext.xml is 60 seconds
-        // timeout specified in ejb-jar.xml is 15 seconds
-        long invokeTime = bean.getInvocationTime();
-        FATHelper.sleep(BUFFER);
-        try {
-            invokeTime = bean.getInvocationTime();
-        } catch (NoSuchEJBException ex) {
-            fail("Expected timeout to be 15 seconds, but bean timed out after " + BUFFER + " ms.");
+        int retry = 0;
+        long secondInvokeTime = 0;
+        StatefulTimeoutXMLOverrideExtFileBean bean = null;
+        while (bean == null) {
+            bean = (StatefulTimeoutXMLOverrideExtFileBean) new InitialContext().lookup(JNDI_XML_OVERRIDE_EXT_FILE);
+            // timeout specified in ibm-ejb-jar-ext.xml is 60 seconds
+            // timeout specified in ejb-jar.xml is 15 seconds
+            long firstInvokeTime = bean.getInvocationTime();
+            FATHelper.sleep(BUFFER);
+            try {
+                secondInvokeTime = bean.getInvocationTime();
+            } catch (NoSuchEJBException ex) {
+                long elapsedTime = System.currentTimeMillis() - firstInvokeTime;
+                if ((retry < 3) && (elapsedTime > USER_TIMEOUT)) {
+                    svLogger.info("Retry - Slow system; elapsed time to call method > timeout : " + elapsedTime + " > " + USER_TIMEOUT);
+                    retry++;
+                    bean = null;
+                } else {
+                    fail("Expected timeout to be 15 seconds, but bean timed out after " + BUFFER + " ms.");
+                }
+            }
         }
 
         FATHelper.sleep(USER_TIMEOUT + BUFFER);
         try {
-            long elapsedTime = bean.getInvocationTime() - invokeTime;
+            long elapsedTime = bean.getInvocationTime() - secondInvokeTime;
             fail("Timeout was " + USER_TIMEOUT + " ms, but bean did not time out after sleeping " + elapsedTime + " ms.");
         } catch (NoSuchEJBException ex) {
             svLogger.info("Caught expected NoSuchEJBException - PASS");

